read-fonts/src/lib.rs

```diff
@@ -71,6 +71,86 @@ pub(crate) mod codegen_prelude {
 
 include!("../generated/font.rs");
 
+#[derive(Clone)]
+/// Reference to the content of a font or font collection file.
+pub enum FileRef<'a> {
+    /// A single font.
+    Font(FontRef<'a>),
+    /// A collection of fonts.
+    Collection(CollectionRef<'a>),
+}
+
+impl<'a> FileRef<'a> {
+    pub fn new(data: FontData<'a>) -> Result<Self, ReadError> {
+        Ok(if let Ok(collection) = CollectionRef::new(data) {
+            Self::Collection(collection)
+        } else {
+            Self::Font(FontRef::new(data)?)
+        })
+    }
+
+    pub fn fonts(&self) -> impl Iterator<Item = Result<FontRef<'a>, ReadError>> + 'a + Clone {
+        let copy = self.clone();
+        let len = match self {
+            Self::Collection(collection) => collection.len(),
+            _ => 1,
+        };
+        (0..len).filter_map(move |ix| match &copy {
+            Self::Font(font) => Some(Ok(font.clone())),
+            Self::Collection(collection) => collection.get(ix),
+        })
+    }
+}
+
+const TTC_HEADER_TAG: Tag = Tag::new(b"ttcf");
+
+/// Reference to the content of a font collection file.
+#[derive(Clone)]
+pub struct CollectionRef<'a> {
+    data: FontData<'a>,
+    header: TTCHeader<'a>,
+}
+
+impl<'a> CollectionRef<'a> {
+    pub fn new(data: FontData<'a>) -> Result<Self, ReadError> {
+        let header = TTCHeader::read(data)?;
+        if header.ttc_tag() != TTC_HEADER_TAG {
+            Err(ReadError::InvalidTtc(header.ttc_tag()))
+        } else {
+            Ok(Self { data, header })
+        }
+    }
+
+    pub fn len(&self) -> u32 {
+        self.header.num_fonts()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    pub fn get(&self, index: u32) -> Option<Result<FontRef<'a>, ReadError>> {
+        if index >= self.len() {
+            return None;
+        }
+        let offset = self
+            .header
+            .table_directory_offsets()
+            .get(index as usize)?
+            .get() as usize;
+        match TableDirectory::read(self.data.slice(offset..)?) {
+            Ok(table_directory) => Some(FontRef::with_table_directory(self.data, table_directory)),
+            _ => None,
+        }
```

Comment:
Could we make the `table_directory_offsets` actually be offsets, and then implement `FontRead` for `FontRef`? 

Also I'm confused by the `None` branch at the bottom; if we can't read the data we should be returning `Some(Err(_))`, I think.

Reply:
Table offsets in font collections are based at the root of the file and not the table directory, so the default offset resolution stuff might not work here?

Good catch on the branch. I saw that but forgot to fix it.

Reply:
The root of the file is also the location of the ttc header, right? So if the header is resolving these offsets it should work.

Alternatively if you have an offset you can always provide your own data, like:

```rust
let some_offset = my_table.some_offset();
let resolved = some_offset.resolve(my_font_data);
```

which still might be easier than manually indexing etc.

Reply:
Ah, yes. I can change that. We just can’t implement `FontRead` for `FontRef` in the case of collections because the base of data and table_directory don’t align in this case.

Reply:
Those offsets are relative to the start of the file. This is so that fonts in a collection can share tables (typically `cmap`). I was actually pleasantly surprised that `FontRef` contained the data field because I would have add to add it for collections to work.